pkg/logging/fluentd_logger.go

```diff
@@ -110,31 +110,37 @@ func (f *FluentdLogger) Process(_ string, config *logging.Config) error {
 	if err != nil {
 		return fmt.Errorf("failed to create fluent client: %w", err)
 	}
+	f.fluentClient = fluentClient
+	f.config = config
+	return nil
+}
+func (f *FluentdLogger) Process(stdout <-chan string, stderr <-chan string) error {
```

Comment:
@Zheaoli please check this answer from [stackoverflow](https://stackoverflow.com/a/13668275)  . seems like no good reason to use channel here

Reply:
Thanks for the tips, I have already read the link you attach to the comment.

The answer from StackOverflow seems to have the same idea as mine.

> Channels are a good fit for communicating between goroutines.

Go back to this PR, please take a look at `loggingProcessAdapter` function here https://github.com/containerd/nerdctl/pull/1519/files#diff-eb19031b20f577302a6bfbcbb42ee60f37256f9ed2044d905a7aa681309bca69R142, you will find that I split scan log message from container's stdout/stderr and process the log message into different goroutine. https://github.com/containerd/nerdctl/pull/1519/files#diff-eb19031b20f577302a6bfbcbb42ee60f37256f9ed2044d905a7aa681309bca69R150-R164

You may wonder why I chose to split this code into different goroutines. Please look at #1470 (still WIP, will update after this PR has been merged); in #1470, I want to allow people to save logs to a local JSON file when they choose fluentd/syslog driver to make debugging easier. If I want to implement this feature, I think it's an excellent choice to use different goroutines to write to the fluentd/syslog and a local JSON file. In this circumstance, I think using a channel is a better way to fit the best practice in Golang.

Reply:
@fahedouch ping~

Reply:
@Zheaoli 
Ok I got the intentation of this refacto!  but  `io` is a good fit too.

Now coming back to  `io`  **vs**  `channel` :

> Channels are a good fit for communicating between goroutines.  

this above sentence is very generic, what is important is the next of the comment which explain what the guy means by `good fit`:  

> When a program does something simple, like for example reading stdin, doing something with the stream and outputting the result to stdout - then using a channel is an overkill, unnecessarily hurting performance.
there's no good reason to model simple operations, like those of io.Reader or io.Writer using channels, respective to have a channel based method set (API).

Looking on your code we are not doing more than playing with stream and outputting it somewhere. So the question here, do we really need to hurt performance by using channel ? 

Other thing worth noting is that by design containerd[ provider io interface for stdout and stdin](https://github.com/containerd/containerd/blob/v1.6.10/runtime/v2/logging/logging.go#L25) So why moving to channels if `io` cloud fit the need

Reply:
> Other thing worth noting is that by design containerd[ provider io interface for stdout and stdin](https://github.com/containerd/containerd/blob/v1.6.10/runtime/v2/logging/logging.go#L25) So why moving to channels if io cloud fit the need

The problem is the containerd's stdou/stderr `io.Reader` is seekable, which means the same position in the IO stream can be read just once.

> You may wonder why I chose to split this code into different goroutines. Please look at https://github.com/containerd/nerdctl/pull/1470 (still WIP, will update after this PR has been merged); in https://github.com/containerd/nerdctl/pull/1470, I want to allow people to save logs to a local JSON file when they choose fluentd/syslog driver to make debugging easier. If I want to implement this feature, I think it's an excellent choice to use different goroutines to write to the fluentd/syslog and a local JSON file. In this circumstance, I think using a channel is a better way to fit the best practice in Golang.

In the up circumstance, two loggers need to read the same message twice. So we need different buffers for logger instances and put the message we read from stdout/stderr in the buffer. For the buffer, I think the channel would be a good choice.

Reply:
makes sense